File: src/protocol/util/tokenize.cpp
/*
 * Copyright (c) 2011 Sveriges Television AB <info@casparcg.com>
 *
 * This file is part of CasparCG (www.casparcg.com).
 *
 * CasparCG is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CasparCG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CasparCG. If not, see <http://www.gnu.org/licenses/>.
 */

#include "tokenize.h"

namespace caspar { namespace IO {

std::size_t tokenize(const std::wstring& message, std::list<std::wstring>& pTokenVector)
{
    // split on whitespace but keep strings within quotationmarks
    // treat \ as the start of an escape-sequence: the following char will indicate what to actually put in the
    // string

    std::wstring currentToken;

    bool inQuote        = false;
    int  inParamList    = 0;
    bool getSpecialCode = false;

    for (unsigned int charIndex = 0; charIndex < message.size(); ++charIndex) {
        if (getSpecialCode) {
            // insert code-handling here
            switch (message[charIndex]) {
                case L'\\':
                    currentToken += L"\\";
                    break;
                case L'\"':
                    currentToken += L"\"";
                    break;
                case L'n':
                    currentToken += L"\n";
                    break;
                default:
                    break;
            };
            getSpecialCode = false;
            continue;
        }

        if (message[charIndex] == L'\\') {
            getSpecialCode = true;
            continue;
        }

        if (message[charIndex] == L' ' && inQuote == false && inParamList == 0) {
            if (!currentToken.empty()) {
                pTokenVector.push_back(currentToken);
                currentToken.clear();
            }
            continue;
        } else if (!inQuote && message[charIndex] == L'(') {
            inParamList++;
            // continue;
        } else if (!inQuote && message[charIndex] == L')') {
            inParamList--;
            if (inParamList == 0) {
                currentToken += message[charIndex];
                pTokenVector.push_back(currentToken);
                currentToken.clear();
                continue;
            }
            // continue;
        } else if (message[charIndex] == L'\"') {
            inQuote = !inQuote;

            if (inParamList == 0) {
                if (!inQuote) {
                    pTokenVector.push_back(currentToken);
                    currentToken.clear();
                }
                continue;
            }
        }

        currentToken += message[charIndex];
    }

    if (!currentToken.empty()) {
        pTokenVector.push_back(currentToken);
        currentToken.clear();
    }

    return pTokenVector.size();
}

}} // namespace caspar::IO

File: node/src/tokenize.ts
export function tokenize(message: string, pTokenVector: string[]): number {
    // split on whitespace but keep strings within quotationmarks
    // treat \ as the start of an escape-sequence: the following char will indicate what to actually put in the
    // string

    let currentToken: string = "";

    let inQuote = false;
    let inParamList = 0;
    let getSpecialCode = false;

    for (const messageChar of message) {
        if (getSpecialCode) {
            // insert code-handling here
            switch (messageChar) {
                case "\\":
                    currentToken += "\\";
                    break;
                case '"':
                    currentToken += '"';
                    break;
                case "n":
                    currentToken += "\n";
                    break;
                default:
                    break;
            }
            getSpecialCode = false;
            continue;
        }

        if (messageChar == "\\") {
            getSpecialCode = true;
            continue;
        }

        if (messageChar == " " && !inQuote && inParamList == 0) {
            if (currentToken.length > 0) {
                pTokenVector.push(currentToken);
                currentToken = "";
            }
            continue;
        } else if (!inQuote && messageChar == "(") {
            inParamList++;
            // continue;
        } else if (!inQuote && messageChar == ")") {
            inParamList--;
            if (inParamList == 0) {
                currentToken += messageChar;
                pTokenVector.push(currentToken);
                currentToken = "";
                continue;
            }
            // continue;
        } else if (messageChar == '"') {
            inQuote = !inQuote;

            if (inParamList == 0) {
                if (!inQuote) {
                    pTokenVector.push(currentToken);
                    currentToken = "";
                }
                continue;
            }
        }

        currentToken += messageChar;
    }

    if (currentToken.length > 0) {
        pTokenVector.push(currentToken);
        // currentToken = "";
    }

    return pTokenVector.length;
}
